### PR TITLE
Throw InvalidOperationException on RuntimeMethodHandle.GetFunctionPointer() for generic methods

### LINKED ISSUE #101664: `RuntimeMethodHandle.GetFunctionPointer()` should check if generic method is constructed
### Description

Calling `GetFunctionPointer()` on a non-constructed generic method throws `InvalidProgramException: Common Language Runtime detected an invalid program`. A more helpful error message could be added.

### Reproduction Steps

```cs
unsafe
{
    var func = (delegate*<void>) typeof(Foo)
        .GetMethod(nameof(Foo.Func))!
        .MethodHandle
        .GetFunctionPointer();
    
    func();
}

static class Foo
{
    public static void Func<T>()
    {
    }
}
```

### Expected behavior

User friendly exception is thrown.

### Actual behavior

`InvalidProgramException` is thrown.

```cs
Unhandled exception. System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at System.RuntimeMethodHandle.GetFunctionPointer() in /runtime/src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:line 829
   at Program.<Main>$(String[] args)
```

### Regression?

No

### Known Workarounds

Not applicable

### Configuration

.NET 8 and 9
linux-x64

### Other information

If the type itself is a generic type it throws `System.InvalidOperationException: Could not execute the method because either the method itself or the containing type is not fully instantiated.`. I think that's the correct exception to be thrown. 

```cs
unsafe
{
    var func = (delegate*<void>) typeof(Foo<>)
        .GetMethod("Func")!
        .MethodHandle
        .GetFunctionPointer();
    
    func();
}

static class Foo<T>
{
    public static void Func()
    {
    }
}
```

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/101664

On CoreClr, the previous exception was InvalidProgramException. On Mono, the call caused a crash.

### CODE REVIEW

**Review by steveisok — APPROVED** [review]

**Comment by lambdageek** [comment]
@steveharter what should this do:

```csharp
public class G<T>
{
    public static T Method(T x) => x;
}

public class Program
{
    public static void Main()
    {
        var mi = typeof(G<>).GetMethod("Method", BindingFlags.Public|BindingFlags.Static);
        var h = mi.MethodHandle;
        var fp = h.GetFunctionPointer();
        Console.WriteLine(fp);
    }
}  
```

On CoreCLR this works (returns some kind of non-null value for `fp`.
On Mono interp it works too.
On Mono JIT it crashes:

```
* Assertion at /Users/runner/work/1/s/src/mono/mono/mini/mini-generic-sharing.c:3368, condition `!mono_class_is_gtd (class_vtable->klass)' not met
```

(same behavior on .NET 9 preview 6 and .NET 8, across both runtimes)

File: src/mono/mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -2938,6 +2938,11 @@ mono_jit_compile_method_jit_only (MonoMethod *method, MonoError *error)
 static gpointer
 get_ftnptr_for_method (MonoMethod *method, gboolean need_unbox, MonoError *error)
 {
+	if (method->is_generic) {
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-runtime.c:2941]: if we want to also rule out methods from a generic type definition, then you may want to add `|| mono_class_is_gtd (method->klass)`

Comment (steveharter) [reply src/mono/mono/mini/mini-runtime.c:2941]: Yes we should throw if the method is not callable meaning a "true" generic method and a generic method that uses the type parameters from the owning class. It seems like `IsGenericMethod()` should return `true` even if the type parameters come from the owning class, but that's a different problem...

I'll make another pass here. Thanks

**Review by lambdageek — APPROVED** [review]

File: src/coreclr/vm/method.cpp [hunk 2]
```diff
@@ -2011,10 +2011,9 @@ PCODE MethodDesc::TryGetMultiCallableAddrOfCode(CORINFO_ACCESS_FLAGS accessFlags
     }
     CONTRACTL_END
 
-    if (IsGenericMethodDefinition())
+    if (IsGenericMethodDefinition() || ContainsGenericVariables())
```
Comment (lambdageek) [inline src/coreclr/vm/method.cpp:2014]: nit:  `ContainsGenericVariables` includes a check for `IsGenericMethodDefinition`

Comment (buyaa-n) [reply src/coreclr/vm/method.cpp:2014]:
```suggestion
    if (ContainsGenericVariables())
```

**Review by buyaa-n — APPROVED** [review]

**Comment by buyaa-n** [comment]
Merging as @steveharter is out

**Pull request merged**